rewrite-java/src/main/java/org/openrewrite/java/trait/Block.java

```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright 2024 the original author or authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.java.trait;
+
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import org.jspecify.annotations.Nullable;
+import org.openrewrite.Cursor;
+import org.openrewrite.Tree;
+import org.openrewrite.TreeVisitor;
+import org.openrewrite.internal.ListUtils;
+import org.openrewrite.java.JavaVisitor;
+import org.openrewrite.java.marker.ImplicitReturn;
+import org.openrewrite.java.tree.*;
+import org.openrewrite.trait.SimpleTraitMatcher;
+import org.openrewrite.trait.Trait;
+import org.openrewrite.trait.VisitFunction2;
+
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import static java.util.Collections.emptyList;
+
+@RequiredArgsConstructor
+public class Block implements Trait<J.Block> {
+    @Getter
+    private final Cursor cursor;
+
+    public Block filterStatements(Predicate<Statement> predicate) {
+        return filterStatements(predicate, (J.Return::withExpression));
+    }
+
+    public Block mapStatements(Function<Statement, @Nullable Statement> mapper) {
+        return mapStatements(mapper, (J.Return::withExpression));
+    }
+
+    public Block filterStatements(Predicate<Statement> predicate, BiFunction<J.Return, Expression, J.Return> returnMapper) {
+        return mapStatements(statement -> predicate.test(statement) ? statement : null, returnMapper);
+    }
```

Comment:
The name `filter` doesn't quite fit since this also maps.
Feels a bit redundant with `mapStatements()`. You can already filter by returning `null` from a mapping.
Since we have the "return `null` means remove" convention already `filter` is a subset of `map`. So for that matter, do we need a separate `filterStatements` at all or is a single `map` method sufficient for everything? But the same could be said about ListUtils and we have `filter` methods there so maybe user convenience justifies the redundancy.

What if these methods returned a `Stream<Statement>` and this exposed a `Collector` to collect a `Stream<Statement>` back into a block?

Reply:
So are we trying to be a bit too creative here? What if we just did a single method like `List<Statement> withStatements(List<Statement> statements)` where we pass in the original list from `J.Block`, the user can update it however they see fit -- likely using `ListUtils *` -- then pass us back the updated list which we move or remove comments in relationship to the two lists. This would support all three use cases and put the handling of aspects such as return statements squarely in the author's hands. The `with*` method here feels fine as well given we're just relaying the state back and forth from the LST, so there's existing precedent for that pattern with other `Trait` implementations.